src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/commissioner/thread/internal/BorderAgentDiscoverer.java

```diff
@@ -0,0 +1,228 @@
+
+    executor.submit(
+        () -> {
+          while (true) {
+            if (!isResolvingService.get()) {
+              NsdServiceInfo serviceInfo = unresolvedServices.take();
+
+              isResolvingService.set(true);
+              nsdManager.resolveService(serviceInfo, listener);
+            }
+          }
+        });
+  }
+
+  private void stopResolver() {
+    if (!executor.isTerminated()) {
+      executor.shutdownNow();
+    }
+    isResolvingService.set(false);
+    unresolvedServices.clear();
+  }
+
+  public void stop() {
+    if (!isScanning) {
+      Log.w(TAG, "the Border Agent discoverer has already been stopped!");
+      return;
+    }
+
+    nsdManager.stopServiceDiscovery(this);
+    stopResolver();
+
+    if (wifiMulticastLock != null) {
+      wifiMulticastLock.release();
+    }
+
+    isScanning = false;
+  }
+
+  @Override
+  public void onDiscoveryStarted(String serviceType) {
+    Log.d(TAG, "start discovering Border Agent");
+  }
+
+  @Override
+  public void onDiscoveryStopped(String serviceType) {
+    Log.d(TAG, "stop discovering Border Agent");
+  }
+
+  @Override
+  public void onServiceFound(NsdServiceInfo nsdServiceInfo) {
+    Log.d(TAG, "a Border Agent service found");
+
+    unresolvedServices.offer(nsdServiceInfo);
+  }
+
+  @Override
+  public void onServiceLost(NsdServiceInfo nsdServiceInfo) {
+    String discriminator = getBorderAgentDiscriminator(nsdServiceInfo);
+    if (discriminator != null) {
+      Log.d(TAG, "a Border Agent service is gone");
+      borderAgentListener.onBorderAgentLost(discriminator);
+    }
+  }
+
+  @Override
+  public void onStartDiscoveryFailed(String serviceType, int errorCode) {
+    Log.d(TAG, "start discovering Border Agent failed: " + errorCode);
+  }
+
+  @Override
+  public void onStopDiscoveryFailed(String serviceType, int errorCode) {
+    Log.d(TAG, "stop discovering Border Agent failed: " + errorCode);
+  }
+
+  private BorderAgentInfo getBorderAgentInfo(NsdServiceInfo serviceInfo) {
+    Map<String, byte[]> attrs = serviceInfo.getAttributes();
+
+    // Use the host address as default discriminator.
+    String discriminator = serviceInfo.getHost().getHostAddress();
+
+    if (attrs.containsKey(KEY_DISCRIMINATOR)) {
+      discriminator = new String(attrs.get(KEY_DISCRIMINATOR));
+    }
+
+    if (!attrs.containsKey(KEY_NETWORK_NAME) || !attrs.containsKey(KEY_EXTENDED_PAN_ID)) {
+      return null;
+    }
+
+    return new BorderAgentInfo(
+        discriminator,
+        new String(attrs.get(KEY_NETWORK_NAME)),
+        attrs.get(KEY_EXTENDED_PAN_ID),
+        serviceInfo.getHost(),
+        serviceInfo.getPort());
+  }
+
+  private String getBorderAgentDiscriminator(NsdServiceInfo serviceInfo) {
+    Map<String, byte[]> attrs = serviceInfo.getAttributes();
+
+    // Use the host address as default discriminator.
+    String discriminator = null;
+
+    if (serviceInfo.getHost() != null) {
+      discriminator = serviceInfo.getHost().getHostAddress();
```

Comment:
should there be some ordering/else since the attrs check would clobber this?

something like:

```
if (attrs.containsKey(KEY_DISCRIMINATOR)) {
  return new String(attrs.get(KEY_DISCRIMINATOR));
}

if (serviceInfo.getHost() != null) {
  return serviceInfo.getHost().getHostAddress();
}

return null;
```